### ISSUE TITLE
CsWin32 analyzer slows Visual Studio 2019 editor down to a crawl

### ISSUE THREAD

**Issue by BasTossings** [microsoft/CsWin32#244]
## Actual behavior

As soon as I add a reference to `Microsoft.Windows.CsWin32` to my project, the Visual Studio 2019 editor slows down to a crawl to the point of barely being usable anymore. The editor lags behing as I type, autocomplete takes several seconds (3~10) to pop up. Code color coding and error underlining is delayed by the same amount.

The moment I remove the reference and hit save, it's back to normal again and editing, autocomplete and code coloring and error underlining is again instantaneous.

## Expected behavior

I'd expect the code generator/analyzer to perform better. I see no reason it should be so slow.

## Repro steps

1. `NativeMethods.txt` content:
```
// Misc

FILE_ACCESS_FLAGS
STATFLAG
STGTY

// General windowing

SetWindowsHookEx
GetCurrentThreadId
CallNextHookEx
GetClientRect
RegisterClassEx
CreateWindowEx
ShowWindow
UpdateWindow
DefWindowProc
ValidateRect
LoadCursor
WM_PAINT
WM_SIZE

// WIC

IWICImagingFactory
IWICBitmapSource

// DirectWrite

DWriteCreateFactory
IDWriteFactory

// Direct2D

D2D1CreateFactory
ID2D1Factory
ID2D1RenderTarget
D2D1_RENDER_TARGET_PROPERTIES
D2D1_HWND_RENDER_TARGET_PROPERTIES
D2D_RECT_F
D2D1_ROUNDED_RECT
D2D1_ELLIPSE
D2D1_FILL_MODE
D2D_MATRIX_3X2_F
D2D1_STROKE_STYLE_PROPERTIES
ID2D1RectangleGeometry
CWPSTRUCT
```

2. `NativeMethods.json` content (if present):
```json
{
  "$schema": "https://aka.ms/CsWin32.schema.json",
  "public": false,
  "allowMarshaling": true
}
```

3. Any of your own code that should be shared?

My solution consists of 74 .NET 5 c# projects

### Context

- CsWin32 version: `0.1.422-beta`
- Win32Metadata version: `10.0.19041.5-preview.138`
- Target Framework: `.NET 5`
- `LangVersion`: n/a

**Comment by bastos80** [comment]
On closer inspection I suspect that the `SourceGenerator`'s Execute is called on every code change in the editor, which would explain the slow down. 

My suggestion would be to cache the `compilationUnits` obtained at line 173 of `SourceGenerator.cs` and only regenerate them when either the contents of `NativeMethods.txt` or `NativeMethods.json` changes (Likely just checking the modified time would be sufficient).

**Comment by sharwell** [comment]
This issue is caused by failing to reuse [cached outputs](https://github.com/dotnet/roslyn/blob/81c70ba8a741b1d9f93edce114fe5e6bd1ebc99c/src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/CachingSourceGenerator.cs#L65-L79) when the [input hasn't changed](https://github.com/dotnet/roslyn/blob/81c70ba8a741b1d9f93edce114fe5e6bd1ebc99c/src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/CachingSourceGenerator.cs#L50-L57). If you prefer, you can use [`SoftReference<T>`](https://github.com/dotnet/roslyn/pull/52330) instead of the Large Object Heap for avoiding premature collection of the cached value.

**Comment by bastos80** [comment]
@AArnott ?

**Comment by AArnott** [comment]
@sharwell Where is this reusing outputs documented? I had no idea that was a thing.

**Comment by AArnott** [comment]
This will largely be a dupe of #38

**Comment by AraHaan** [comment]
> This issue is caused by failing to reuse [cached outputs](https://github.com/dotnet/roslyn/blob/81c70ba8a741b1d9f93edce114fe5e6bd1ebc99c/src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/CachingSourceGenerator.cs#L65-L79) when the [input hasn't changed](https://github.com/dotnet/roslyn/blob/81c70ba8a741b1d9f93edce114fe5e6bd1ebc99c/src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/CachingSourceGenerator.cs#L50-L57). If you prefer, you can use [`SoftReference<T>`](https://github.com/dotnet/roslyn/pull/52330) instead of the Large Object Heap for avoiding premature collection of the cached value.

Not just that, it could be contributed with the usage of YamlDotNet in ``Docs`` class which is used in ``Generator`` to return the docs in the generated code. I am filing an issue in the metadata repository to embed the docs in the winmd files in a way that all languages can extract && make use of them without too much effort or performance issues.

**Comment by AArnott** [comment]
Thanks, although we didn't really need that extra issue. I'm doing the work of moving the docs to the win32metadata repo as part of the issues we already had. :)

**Comment by AArnott** [comment]
Several perf fixes have been made, Can you try 0.1.505-beta or later (from [our CI feed](https://dev.azure.com/azure-public/winsdk/_packaging?_a=package&feed=CI&package=Microsoft.Windows.CsWin32&protocolType=NuGet&version=0.1.505-beta)) and let us know how it's looking?

We are not caching the actual generated output yet. We could potentially do that, but I'm not sure how effective it would be because in fact the inputs span more than just NativeMethods files. Because we allow you to define or reference your own types for ours, any change you make (manipulate references or code changes) could potentially alter the generated output.

**Comment by AArnott** [comment]
@BasTossings you may want to review [this comment](https://github.com/microsoft/CsWin32/discussions/591#discussioncomment-3199262), as apparently a VS extension has been known to interact negatively with source generators and needs a fix. 
I'm going to close this issue as fixed given the perf work we've already done. But I'll open a new issue to track future perf opportunities as well.
